### PR TITLE
Fixed mailcollector with multibyte filename (#8495)

### PR DESCRIPTION
**Pull request by edgardmessias**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #8495

https://github.com/laminas/laminas-mail/pull/128

### CODE REVIEW

**Comment by trasher** [comment]
Sounds like the fix is correct, thank you :+1: 
Anyways, we need to add the case in GLPI tests, by adding a .eml file that permit to reproduce the initial issue.

**Comment by tomolimo** [comment]
Hello @edgardmessias 

Here is the eml file,
[22-multibyte_attachment_name.zip](https://github.com/glpi-project/glpi/files/5819336/22-multibyte_attachment_name.zip)

Thank you
Regards,
Tomolimo

**Comment by tomolimo** [comment]
Hello @edgardmessias 

I made a test with your solution: it's OK, and I think it's better than mine :smiley: 

However, could you add a test to be sure that `file_put_contents($path.$filename, $contents)` will be accepted by OS?
Who knows if the attachment name has invalid chars for the OS hosting GLPI...

```
         // try to save file to disk, to be sure file name is valid (= accepted by OS)
         $ret = @file_put_contents($path.$filename, $contents);
         if ($ret === false) {
            // if not, then create a valid file name
            $ext = pathinfo($filename, PATHINFO_EXTENSION);
            $strip = ["*", "\\", "|", ":", "\"", "<", ">", "/", "?"];
            $filename = "doc_{$subpart}_(renamed)_"
               . trim(str_replace($strip, "", pathinfo($filename, PATHINFO_FILENAME)))
               . ".$ext";
            // retry to save the file
            $ret = file_put_contents($path.$filename, $contents);
         }
         if ($ret) {
```

Thank you
Regards,

Tomolimo

**Comment by tomolimo** [comment]
Hello @edgardmessias 
Here is the eml for the bad chars for OS:
[23-bad_attachment_name_for_OS.zip](https://github.com/glpi-project/glpi/files/5820727/23-bad_attachment_name_for_OS.zip)
Thank you,
Regards,
Tomolimo

**Comment by edgardmessias** [comment]
Hello @tomolimo .
Big thanks.

As possible, I will write the unit tests

**Comment by trasher** [comment]
Thank you very much @edgardmessias! Please feel free to ping us if you have issues with tests.

**Comment by glensc** [comment]
btw, you may want to use this package for filename normalizing:
- https://packagist.org/packages/glen/filename-normalizer

**Comment by edgardmessias** [comment]
@trasher @tomolimo , Could you help me to fix the test? In my local environment is working fine.

**Comment by trasher** [comment]
Hi @edgardmessias I'll try to take a look today to understand what's going wrong

File: inc/mailcollector.class.php
Comment (trasher) [inline inc/mailcollector.class.php:1513]: Is that strictly required? I do not like `@` usage

Comment (cedric-anne) [reply inc/mailcollector.class.php:1513]: If there is an error triggered by laminas-mail on some emails, we should, at least, open an issue on their side.

I do not like suppressing errors, especially when the called method can have many unpredictible reasons to trigger an error. Indeed, you may want to suppress a specific error that may not affect us, but you will also mute all other errors related to any email parsing.

**Comment by trasher** [comment]
Error in tests is the following:
```
'2021-01-18 07:46:50 [@976e2e5af287]
Message is invalid: The input is not a valid email address. Use the basic format local-part@hostname<br/>2021-01-18 07:46:50 [@976e2e5af287]
Error during message parsing: Header with Name date or date not found<br/>2021-01-18 07:46:50 [@976e2e5af287]
Error during message parsing: str_split(): Argument #2 ($length) must be greater than 0<br/>'
```

Found it reading the `mailgate.log` file from tests

**Comment by tomolimo** [comment]
Hello @edgardmessias @trasher 
On my environment I got the same kind or problem, the errors counter is not OK.
I believe, this is due to one of the emails that can or cannot (depending on the mail server) serves it.
I think (but not sure) that it's coming from the chat email (eml file name is: `05-empty-from.eml`)
Regards,
Thank you,
Tomolimo

File: inc/mailcollector.class.php
Comment (edgardmessias) [reply inc/mailcollector.class.php:1513]: The main problem in laminas-mails is `str_split(): Argument #2 ($length) must be greater than 0`.
I openned a PR to resolve it: https://github.com/laminas/laminas-mail/pull/128

Comment (trasher) [reply inc/mailcollector.class.php:1513]: OK, so we'll have to wait for new laminas-mail release - or at least for that PR to be approved

**Review by stonebuzz — CHANGES_REQUESTED** [review]
this part
```php
         // Try to get filename from Content-Type header
         if (empty($filename)) {
            $filename = $content_type_header->getParameter('name') ?? '';
         }
```

need to be moved after the attempt to convert the filename with encoding value (espaclially with base64encoding) 
Otherwise the name of the file becomes the title of the email.

You can try with this file 

[DUC_SING WAH PAN (MA Chan Ho Ki).pdf](https://github.com/glpi-project/glpi/files/5848783/DUC_SING.WAH.PAN.MA.Chan.Ho.Ki.pdf)

Done by latest commit

**Comment by stonebuzz** [comment]
Sorry @edgardmessias I have force-push to you PR :confused:

**Comment by edgardmessias** [comment]
@stonebuzz , don't worry

File: .github/workflows/ci.yml
Comment (cedric-anne) [inline .github/workflows/ci.yml:230]: I put tests on the top, but this should be reverted before merge.

**Review by cedric-anne — COMMENTED** [review]
@edgardmessias 

I applied your patch and adapted a bit tested emails, but I think there is still a bug in laminas component, in the method that returns the filename value. As I changed manually tested mails headers, Iit may be due to a bad manipulation, but filenames are correct when I open them with Thunderbird.

You can check test suite output.

File: .github/workflows/ci.yml
Comment (edgardmessias) [reply .github/workflows/ci.yml:230]: Yes, I looking here what is the problem

**Comment by edgardmessias** [comment]
@cedric-anne , I think it is a good idea to append this code for bad header:

```php
if (strncmp($filename, '=?', 2) === 0) {
   $filename = preg_replace('/\?\=\=\?[\w-]+\?\w\?/', '', $filename);
   var_dump($filename);
   $filename = \Laminas\Mail\Header\HeaderWrap::mimeDecodeValue($filename);
   var_dump($filename);
}
```

File: tests/imap/MailCollector.php
Comment (cedric-anne) [inline tests/imap/MailCollector.php:600]: @edgardmessias 

Did you saw that ?

I guess this is only a side effect of your fix. Indeed, now the space char is located at the beginning or end of a param line, and trimming done in Laminas removes it. Before your fix, this space was certainly located in the middle of second parameter line.

Param lines are extracted with these values:
```
array(2) {
  [0]=>
  string(45) "01-Screenshot-2018-4-12 Observatoire - France"
  [1]=>
  string(21) "très haut débit.png"
}
```

**Review by cedric-anne — APPROVED** [review]

**Comment by edgardmessias** [comment]
@cedric-anne , I fixed the problem with malformed header: https://github.com/laminas/laminas-mail/pull/128/commits/2f81aedcb83de61e00fffd6aee4c8938272a1b24
This solve the problem with previous eml file

**Comment by cedric-anne** [comment]
> @cedric-anne , I think it is a good idea to append this code for bad header:
> 
> ```
> if (strncmp($filename, '=?', 2) === 0) {
>    $filename = preg_replace('/\?\=\=\?[\w-]+\?\w\?/', '', $filename);
>    var_dump($filename);
>    $filename = \Laminas\Mail\Header\HeaderWrap::mimeDecodeValue($filename);
>    var_dump($filename);
> }
> ```

Did you already encounter this case ? Was it due to a malformed email, or to a laminas-mail bug ?

Trying to fix laminas-mail behaviour in our side should only be done if we are not able to propose them a fix. (This has been done to handle RFC5987 format for filenames only because I was not able to find how to fix that on their side).

Trying to fix malformed emails is not always a good idea, it may introduce side effects.

Unless you are able to quickly propose a fix and test(s) email(s) that validates the fix, you should open another issue/PR for this case, to not block merging of current one.

Regards

**Comment by tomolimo** [comment]
Hello @edgardmessias 
image

I have some troubles with the `patch` command which is not available on Windows...
Could it be possible to use agnostic tools?
Thank you
Regards,
Tomolimo

**Comment by cedric-anne** [comment]
> Hello @edgardmessias
> image
> 
> I have some troubles with the `patch` command which is not available on Windows...
> Could it be possible to use agnostic tools?
> Thank you
> Regards,
> Tomolimo

I would love to use agnostic tool for this. Do you know a tool that works both on Windows, Alpine linux, Debian and Fedora ?

**Comment by tomolimo** [comment]
> I would love to use agnostic tool for this. Do you know a tool that works both on Windows, Alpine linux, Debian and Fedora ?

Currently I failed finding a patch tool for Windows...

**Comment by edgardmessias** [comment]
We can use `git apply`

**Comment by cedric-anne** [comment]
> We can use `git apply`

Not sure of this, if dependency was fetched using dist, then there is no git index and patch may not be applied.

**Comment by tomolimo** [comment]
I found a way: I've used the WSL2 to run the `composer install` command

**Comment by edgardmessias** [comment]
I do know if this work, but I will test it:
https://github.com/netresearch/composer-patches-plugin
https://github.com/vaimo/composer-patches

**Review by trasher — APPROVED** [review]

**Pull request merged**
